modules/cudaarithm/src/core.cpp

```diff
@@ -163,8 +163,10 @@ void cv::cuda::flip(InputArray _src, OutputArray _dst, int flipCode, Stream& str
 
     _dst.create(src.size(), src.type());
     GpuMat dst = getOutputMat(_dst, src.size(), src.type(), stream);
+    if (src.data == dst.data && ((src.cols & 1) == 1 || (src.rows & 1) == 1))
+        CV_Error(Error::BadROISize, "In-place version of flip only accepts even width/height");
 
-    if (src.refcount != dst.refcount)
+    if (src.data != dst.data)
```

Comment:
BTW, `refcount` is a pointer to reference counter which is allocated with underlying buffer/data and it is shared between GpuMats.
At least, it is more reliable check in case of image ROIs.

But it doesn't work if `refcount == NULL` (non-managed buffers).

Better check should compare both values:
```
bool isInplace = (src.data == dst.data) || (src.refcount == dst.refcount);
```
(however there are still some non-covered cases)